### PR TITLE
add i2c transaction cue to the linux arch

### PR DESCRIPTION
**Pull request by kirkscheper**
This adds a thread to read/write i2c transactions so that they don't block the main thread. This should address #1500. I currently have a hard sleep in the i2c thread of 5us, let me know if this is appropriate. Of course the transactions will not come in too fast but you also don't want too much latency.

### CODE REVIEW

**Comment by kirkscheper** [comment]
Seems to work. Tested on Bebop 2, with 250us sleep no i2c errors and cpu usage ~3% with 3 i2cs being used.

**Comment by kirkscheper** [comment]
I don't think that all of the mutex locks you mention are strictly necessary but for the sake of safety I added them.

I now split each i2c dev to a different thread and replaced the polling with some signaling. Given that each i2c transaction is initiated with a write, I don't think that select or poll are useful here.

This is untested (even uncompiled... don't have pprz on my desktop) but any comments would be welcome.

File: sw/airborne/arch/linux/mcu_periph/i2c_arch.c [hunk 1]
```diff
@@ -47,39 +126,102 @@ bool i2c_idle(struct i2c_periph *p __attribute__((unused)))
...
-#pragma GCC diagnostic ignored "-Wcast-qual"
 bool i2c_submit(struct i2c_periph *p, struct i2c_transaction *t)
 {
-  int file = (int)p->reg_addr;
+  static uint8_t temp;
+  static pthread_mutex_t *i2c_mutex;
+  static pthread_cond_t *condition;
+
+  // find mutex lock and condition for the appropriate thread
+  for (int i = 0; i < N_I2C; i++){
```
[the remark this answers is not in the document]
Comment (kirkscheper) [reply sw/airborne/arch/linux/mcu_periph/i2c_arch.c]: I didn't know about the init_struct. Much more elegant solution. Updated!

File: sw/airborne/arch/linux/mcu_periph/i2c_arch.c [hunk 2]
```diff
@@ -92,24 +234,31 @@ bool i2c_submit(struct i2c_periph *p, struct i2c_transaction *t)
...
+        t->status = I2CTransSuccess;
       }
       break;
     default:
       break;
+    }
+
+    if(t->status == I2CTransSuccess) {
+      pthread_mutex_lock(&data->mutex);
+      data->p->trans_extract_idx = (data->p->trans_extract_idx + 1) % I2C_TRANSACTION_QUEUE_LEN;
```
[the remark this answers is not in the document]
Comment (kirkscheper) [reply sw/airborne/arch/linux/mcu_periph/i2c_arch.c]: I thought to resend but indeed maybe better to just skip and move on.

**Comment by kirkscheper** [comment]
I had meant it as a default, 'true if not false' type thing but indeed I forgot the default case.

You tested it? All working? I won't have time to test till Monday

**Comment by gautierhattenberger** [comment]
Actually, the problem is not the default case, but as soon as you set to Success, the main code will process the data, so you should set it once the transaction is completed.
Anyway it is tested and working on bebop2 at least.

**Comment by hooperfly** [comment]
Breaks the NPS airframe compilations on  Mac OS X. 

Looks like a `pthread_setname_np` difference between Linux and Mac OS. Probably need some conditional compilation brackets to handle this issue.

Some info on the issue:

Build is broken on OS X because it is only possible to set thread name for the current thread and pthread_setname_np accepts a single argument. This function appeared in 10.6. If that OS version needs to be supported, the best thing that can be done is conditional compilation. A short summary of inconsistencies between different *nix systems can be found here: ​http://stackoverflow.com/questions/2369738/can-i-set-the-name-of-a-thread-in-pthreads-linux/7989973#7989973

**Pull request merged**